[awtrix3] Initial contribution (#18242)
openhab/openhab-addons: bundles/org.openhab.binding.mqtt.awtrixlight/README.md

# MQTT Awtrix 3 Binding

> This binding allows you to control Awtrix 3 (formerly Awtrix Light) LED matrix displays via MQTT. The Awtrix 3 is a customizable 32x8 LED matrix display that can show various information like time, weather, notifications and custom text/graphics.

Reviewer: Each sentence on its own line, please do so for the whole readme.

```suggestion
This binding allows you to control Awtrix 3 (formerly Awtrix Light) LED matrix displays via MQTT.
The Awtrix 3 is a customizable 32x8 LED matrix display that can show various information like time, weather, notifications and custom text/graphics.
```

## Supported Things

This binding supports two types of things:

| Thing Type             | Description                                                                                     |
|------------------------|-------------------------------------------------------------------------------------------------|

> | `awtrixclock` (Bridge) | Represents an Awtrix 3 display device. Acts as a bridge for apps.                               |
> | `awtrixapp`            | Represents an app running on the Awtrix display. Apps can show text, icons, notifications, etc. |

Reviewer: Please stick to the naming convention for thing id's

```suggestion
| `awtrix-clock` (Bridge) | Represents an Awtrix 3 display device. Acts as a bridge for apps.                               |
| `awtrix-app`            | Represents an app running on the Awtrix display. Apps can show text, icons, notifications, etc. |
```

Author: Done

## Channels

### Bridge Channels (`awtrixclock`)

| Channel           | Type                 | Description                                                                                                                                    |
|-------------------|----------------------|------------------------------------------------------------------------------------------------------------------------------------------------|
| `app`             | String               | Currently active app: Will show the name of the app that is currently shown on the display.                                                    |

> | `autoBrightness`  | Switch               | Automatic brightness control: The clock will adjust the display brightness automatically based on ambient light.                               |

Reviewer: The `lower-case-hyphen` naming convention should be used for the channel id's

Author: Done

| `scrollSpeed`        | Number:Dimensionless | Text scrolling speed: Provide as percentage value. The original speed is 100%. Values above 100% will increase the scrolling speed, values below 100% will decrease it. Setting this value to 0 will disable scrolling completely.                                                                                                                     |
| `text`               | String               | Text to display.                                                                                                                            |
| `textCase`           | Number:Dimensionless | Set text case (0=normal, 1=uppercase, 2=lowercase).                                                                                         |
| `textOffset`         | Number:Dimensionless | Text offset position: Horizontal offset of the text in pixels.                                                                              |
| `topText`            | String               | Draws the text on the top of the display.                                                                                                   |

## Full Example

### Things

> ```

Reviewer: Add language to code blocks.
```suggestion
```java
```

Author: Done

Bridge mqtt:broker:myBroker [ host="localhost", port=1883 ]
Bridge mqtt:awtrixclock:myBroker:myAwtrix "Living Room Display" (mqtt:broker:myBroker) [ basetopic="awtrix", appLockTimeout=10, lowBatteryThreshold=25 ] {
    Thing awtrixapp clock "Clock App" [ appname="clock", useButtons=true ]
    Thing awtrixapp weather "Weather App" [ appname="weather" ]
    Thing awtrixapp calendar "Calendar App" [ appname="calendar" ]
    Thing awtrixapp custom "Custom App" [ appname="custom" ]
}
```

### Items

> ```

Reviewer: ```suggestion
```java
```

Author: Done

// Custom App items with advanced features
Switch Custom_Active "Custom App Active" (gAwtrix) { channel="mqtt:awtrixapp:myBroker:myAwtrix:custom:active" }
String Custom_Text "Custom Text" (gAwtrix) { channel="mqtt:awtrixapp:myBroker:myAwtrix:custom:text" }
String Custom_Icon "Custom Icon" (gAwtrix) { channel="mqtt:awtrixapp:myBroker:myAwtrix:custom:icon" }
Color Custom_Color "Text Color" (gAwtrix) { channel="mqtt:awtrixapp:myBroker:myAwtrix:custom:color" }
Color Custom_Background "Background Color" (gAwtrix) { channel="mqtt:awtrixapp:myBroker:myAwtrix:custom:background" }
Number:Dimensionless Custom_ScrollSpeed "Scroll Speed" (gAwtrix) { channel="mqtt:awtrixapp:myBroker:myAwtrix:custom:scrollSpeed" }
Switch Custom_Center "Center Text" (gAwtrix) { channel="mqtt:awtrixapp:myBroker:myAwtrix:custom:center" }
Number:Dimensionless Custom_Progress "Progress [%d %%]" (gAwtrix) { channel="mqtt:awtrixapp:myBroker:myAwtrix:custom:progress" }
Color Custom_ProgressColor "Progress Color" (gAwtrix) { channel="mqtt:awtrixapp:myBroker:myAwtrix:custom:progressColor" }
```

### Sitemap

> ```
> 

Reviewer: ```suggestion
```perl
```

Author: Done

        Switch item=Display_Power
        Slider item=Display_Brightness
        Switch item=Display_Screen
        Switch item=Display_Sound
        Switch item=Display_AutoBrightness
        Text item=Display_Temperature
        Text item=Display_Humidity
        Text item=Display_Battery visibility=[Display_LowBattery==ON]
        Text item=Display_WiFi
        Text item=Display_CurrentApp
    }
    
    Frame label="Clock App" {
        Switch item=Clock_Active
        Text item=Clock_Text
        Colorpicker item=Clock_Color
        Slider item=Clock_Duration
    }
    
    Frame label="Weather App" {
        Switch item=Weather_Active
        Text item=Weather_Text
        Text item=Weather_Icon
        Colorpicker item=Weather_Color
        Switch item=Weather_Rainbow
    }
    
    Frame label="Custom App" {
        Switch item=Custom_Active
        Text item=Custom_Text
        Text item=Custom_Icon
        Colorpicker item=Custom_Color
        Colorpicker item=Custom_Background
        Slider item=Custom_ScrollSpeed
        Switch item=Custom_Center
        Slider item=Custom_Progress
        Colorpicker item=Custom_ProgressColor
    }
}

```

## Discovery

> The binding can automatically discover Awtrix devices that publish their status to the configured MQTT broker. Once a device is discovered, it will appear in the inbox. Default Awtrix apps can also be discovered if `discoverDefaultApps` is enabled on the bridge. This is however not recommended as the default apps cannot be controlled via this binding.

Reviewer: This section should be above the bridge section.

Author: Done

    'color' -> [255,165,0],  // Orange color
    'rainbow' -> true,
    'duration' -> 10
)
awtrixActions.showCustomNotification(
    params,     // Notification parameters
    false,      // hold: Keep notification until manually cleared
    true,       // wakeUp: Wake up from screen saver
    true,       // stack: Add to notification stack
    "Indiana:d=4,o=5,b=250:e,8p,8f,8g,8p,1c6,8p.,d,8p,8e,1f,p.,g,8p,8a,8b,8p,1f6,p,a,8p,8b,2c6,2d6,2e6,e,8p,8f,8g,8p,1c6,p,d6,8p,8e6,1f.6,g,8p,8g,e.6,8p,d6,8p,8g,e.6,8p,d6,8p,8g,f.6,8p,e6,8p,8d6,2c6", // RTTTL sound to play (not both sound and rtttl)
    "alert",    // Sound file to play (not both sound and rtttl)
    false       // loopSound: Loop the sound until manually stopped
)
```

#### Custom Notification Parameters

The action method parameters:

| Parameter | Type | Description |
|-----------|------|-------------|
| `hold` | Boolean | Keep notification until manually cleared |
| `wakeUp` | Boolean | Wake up from screen saver |
| `stack` | Boolean | Add to notification stack |
| `rtttl` | String | RTTTL melody to play |
| `sound` | String | Sound file to play (without extension) |
| `loopSound` | Boolean | Loop the sound |

> | `params` | Map | Notification parameters |

Reviewer: Better to use markdowntable.com or any other markdown tool to align the table.

Author: Done
